### PR TITLE
Fix Node.js version detection for version managers

### PR DESCRIPTION
**Pull request by feifei325**
## Summary
This PR fixes issue wecode-ai/RunVSAgent#43 where Node.js installed via version managers (asdf, volta, nvm, fnm) was not being detected properly.

## Changes
- Enhanced Node.js detection to support multiple version managers
- Added fallback detection using `which`/`where` command
- Included version manager paths in PATH environment variable
- Added dedicated method to build additional search paths for Node.js

## Supported Version Managers
- asdf
- volta
- nvm
- fnm
- Scoop (Windows)
- Homebrew (macOS)

## Testing
The fix attempts multiple detection methods in sequence:
1. Standard PATH lookup using PathEnvironmentVariableUtil
2. Search in common version manager directories
3. Execute `which`/`where` command as fallback

This should resolve the "Node.js version too low" error for users with valid Node.js installations managed by these tools.

### CODE REVIEW

File: jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt [hunk 1]
```diff
@@ -317,9 +317,102 @@ class ExtensionProcessManager : Disposable {
      * Find executable in system path
      */
     private fun findExecutableInPath(name: String): String? {
```
Comment (feifei325) [inline jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt:319]: Consider adding null safety check for `name` parameter to prevent potential NPE.

File: jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt [hunk 2]
```diff
@@ -317,9 +317,102 @@ class ExtensionProcessManager : Disposable {
...
+        // First try the standard path finding
         val nodePath = PathEnvironmentVariableUtil.findExecutableInPathOnAnyOS("node")?.absolutePath
```
Comment (feifei325) [inline jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt:321]: Good improvement: Added null check for nodePath. This ensures fallback methods are tried when standard path detection fails.

File: jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt [hunk 3]
```diff
@@ -317,9 +317,102 @@ class ExtensionProcessManager : Disposable {
...
-        LOG.info("System Node path: $nodePath")
-        return nodePath
+        if (nodePath != null) {
+            LOG.info("System Node path found: $nodePath")
+            return nodePath
+        }
+        
+        // If not found, try additional paths for version managers
+        val additionalPaths = buildAdditionalNodePaths()
```
Comment (feifei325) [inline jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt:328]: Consider caching the result of `buildAdditionalNodePaths()` to avoid repeated computations on multiple calls.

File: jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt [hunk 4]
```diff
@@ -317,9 +317,102 @@ class ExtensionProcessManager : Disposable {
...
+        for (path in additionalPaths) {
+            val nodeFile = File(path, if (SystemInfo.isWindows) "node.exe" else "node")
+            if (nodeFile.exists() && nodeFile.canExecute()) {
```
Comment (feifei325) [inline jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt:331]: Good defensive programming: Checking both `exists()` and `canExecute()` ensures the found node binary is actually usable.

File: jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt [hunk 5]
```diff
@@ -317,9 +317,102 @@ class ExtensionProcessManager : Disposable {
...
+        }
+        
+        // Try to execute which/where command to find node
+        val whichCommand = if (SystemInfo.isWindows) "where" else "which"
+        try {
+            val process = ProcessBuilder(whichCommand, "node").start()
+            val output = process.inputStream.bufferedReader().readText().trim()
+            process.waitFor(2, TimeUnit.SECONDS)
+            
+            if (output.isNotEmpty()) {
```
Comment (feifei325) [inline jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt:344]: The 2-second timeout is reasonable, but consider making it configurable or using a constant for better maintainability.

File: jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt [hunk 6]
```diff
@@ -317,9 +317,102 @@ class ExtensionProcessManager : Disposable {
...
+            }
+        }
+        
+        // Try to execute which/where command to find node
+        val whichCommand = if (SystemInfo.isWindows) "where" else "which"
+        try {
+            val process = ProcessBuilder(whichCommand, "node").start()
+            val output = process.inputStream.bufferedReader().readText().trim()
+            process.waitFor(2, TimeUnit.SECONDS)
+
```
Comment (feifei325) [inline jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt:343]: Consider using `use{}` extension function for automatically closing the InputStream to prevent resource leaks.

File: jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt [hunk 7]
```diff
@@ -317,9 +317,102 @@ class ExtensionProcessManager : Disposable {
...
+                    val trimmedLine = line.trim()
+                    if (trimmedLine.isNotEmpty() && File(trimmedLine).exists()) {
+                        LOG.info("Found Node.js via $whichCommand command: $trimmedLine")
+                        return trimmedLine
+                    }
+                }
+            }
+        } catch (e: Exception) {
+            LOG.debug("Failed to find node via $whichCommand command", e)
+        }
```
Comment (feifei325) [inline jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt:356]: Good error handling: Using debug level for the exception allows troubleshooting without flooding logs in production.

File: jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt [hunk 8]
```diff
@@ -317,9 +317,102 @@ class ExtensionProcessManager : Disposable {
...
+        LOG.warn("Node.js not found in system path")
+        return null
+    }
+    
+    /**
+     * Build additional paths to search for Node.js (for version managers)
+     */
+    private fun buildAdditionalNodePaths(): List<String> {
+        val paths = mutableListOf<String>()
+        val home = System.getProperty("user.home")
```
Comment (feifei325) [inline jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt:367]: Consider adding validation for `home` directory path to handle edge cases where user.home property might be null or invalid.

File: jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt [hunk 9]
```diff
@@ -317,9 +317,102 @@ class ExtensionProcessManager : Disposable {
...
+        
+        // Common paths for version managers
+        if (SystemInfo.isMac || SystemInfo.isLinux) {
+            // asdf paths
```
Comment (feifei325) [inline jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt:371]: Excellent coverage of version managers. The paths for asdf, volta, nvm, and fnm cover the most popular Node.js version managers.

File: jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt [hunk 10]
```diff
@@ -317,9 +317,102 @@ class ExtensionProcessManager : Disposable {
...
+                paths.add("/usr/local/bin")
+            }
+            
+            // Common Linux paths
+            if (SystemInfo.isLinux) {
+                paths.add("/usr/bin")
+                paths.add("/usr/local/bin")
+            }
+        } else if (SystemInfo.isWindows) {
+            // volta paths on Windows
```
Comment (feifei325) [inline jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt:395]: Consider using File.separator or Paths.get() for cross-platform path construction instead of hardcoded backslashes.

File: jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt [hunk 11]
```diff
@@ -317,9 +317,102 @@ class ExtensionProcessManager : Disposable {
...
+            paths.add("$home\\.volta\\bin")
+            
+            // nvm-windows paths
```
Comment (feifei325) [inline jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt:398]: Good addition of null safety check for APPDATA environment variable on Windows systems.

File: jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt [hunk 12]
```diff
@@ -384,23 +477,43 @@ class ExtensionProcessManager : Disposable {
...
                 "C:\\Program Files\\PowerShell\\7",
-                "C:\\Program Files (x86)\\PowerShell\\7"
+                "C:\\Program Files (x86)\\PowerShell\\7",
+                // Version manager paths
+                "$home\\.volta\\bin",
+                "$home\\scoop\\shims",
+                "C:\\Program Files\\nodejs",
+                "C:\\Program Files (x86)\\nodejs"
+            )
+            else -> listOf(
```
Comment (feifei325) [inline jetbrains_plugin/src/main/kotlin/com/sina/weibo/agent/core/ExtensionProcessManager.kt:507]: Excellent improvement: Replacing the empty list for non-Windows/Mac platforms with actual Linux-specific paths significantly improves Node.js detection on Linux systems.

File: jetbrains_plugin/src/test/kotlin/com/sina/weibo/agent/core/ExtensionProcessManagerTest.kt [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+
+    private lateinit var processManager: ExtensionProcessManager
+    private val originalUserHome = System.getProperty("user.home")
+
+    override fun setUp() {
+        super.setUp()
+        processManager = ExtensionProcessManager()
+    }
+
+    override fun tearDown() {
```
Comment (feifei325) [inline jetbrains_plugin/src/test/kotlin/com/sina/weibo/agent/core/ExtensionProcessManagerTest.kt:27]: Good practice: Properly restoring system properties in tearDown to avoid test pollution.

File: jetbrains_plugin/src/test/kotlin/com/sina/weibo/agent/core/ExtensionProcessManagerTest.kt [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+    fun testBuildAdditionalNodePaths_AllPathsAreStrings() {
+        // Test that all returned paths are valid strings
+        val result = invokePrivateMethod("buildAdditionalNodePaths") as List<String>
+        
+        result.forEach { path ->
+            assertNotNull("Path should not be null", path)
+            assertFalse("Path should not be empty", path.isEmpty())
+            assertFalse("Path should not be blank", path.isBlank())
+        }
+    }
```
Comment (feifei325) [inline jetbrains_plugin/src/test/kotlin/com/sina/weibo/agent/core/ExtensionProcessManagerTest.kt:95]: Consider adding more edge case tests such as empty home directory, invalid paths, or permission issues to increase test coverage.

File: jetbrains_plugin/src/test/kotlin/com/sina/weibo/agent/core/ExtensionProcessManagerTest.kt [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+    fun testMacSpecificPaths() {
+        if (SystemInfo.isMac) {
+            val result = invokePrivateMethod("buildAdditionalNodePaths") as List<String>
+            
+            // Should contain Mac-specific paths
+            assertTrue("Should contain homebrew paths on Mac",
+                result.any { it.contains("/opt/homebrew") || it.contains("/usr/local") })
+        }
+    }
+
```
Comment (feifei325) [inline jetbrains_plugin/src/test/kotlin/com/sina/weibo/agent/core/ExtensionProcessManagerTest.kt:184]: The reflection-based testing approach is reasonable, but consider extracting the private methods to package-private visibility or using a testing framework that supports better private method testing.

**Review by feifei325 — COMMENTED** [review]
## Code Review Summary

Overall, this is a well-implemented enhancement that significantly improves Node.js detection for version managers. The code properly handles cross-platform differences and includes comprehensive fallback mechanisms.

### Key Strengths:
- ✅ Comprehensive version manager support (asdf, volta, nvm, fnm)
- ✅ Proper cross-platform handling (Windows, macOS, Linux)
- ✅ Good error handling and logging
- ✅ Comprehensive test coverage
- ✅ Proper fallback mechanisms with `which`/`where` commands

### Areas for Improvement:
- Consider caching computed paths for performance
- Use cross-platform path construction utilities
- Add more edge case tests
- Consider resource management improvements

The implementation follows good software engineering practices and should resolve the Node.js detection issues for users with version managers.
